pkg/imagefs/imagefs.go

```
@@ -0,0 +1,233 @@
+	logrus.Debugf("imagefs: Lstat file %s", name)
+	if fi, err := ifs.FS.Lstat(name); err == nil {
+		return fi, nil
+	}
+	if ifs.files[name] != nil {
+		logrus.Debugf("imagefs: Lstat cached file %s", name)
+		return ifs.files[name], nil
+	}
+	return nil, fs.ErrNotExist
+}
+
+func (ifs *imageFS) Stat(name string) (fs.FileInfo, error) {
+	logrus.Debugf("imagefs: Stat file %s", name)
+	if fi, err := ifs.FS.Stat(name); err == nil {
+		return fi, nil
+	}
+	if ifs.files[name] != nil {
+		logrus.Debugf("imagefs: Stat cached file %s", name)
+		return ifs.files[name], nil
+	}
+	return nil, fs.ErrNotExist
+}
+
+func (ifs *imageFS) ReadDir(name string) ([]fs.DirEntry, error) {
+	logrus.Debugf("imagefs: Reading directory %s", name)
+	if de, err := ifs.FS.ReadDir(name); err == nil {
+		return de, nil
+	}
+	for dir, d := range ifs.dirs {
+		if ok, err := filepath.Match(name, dir); ok && err == nil {
+			logrus.Debugf("imagefs: Reading cached directory %s", name)
+			return d.entry, nil
+		}
+	}
+	return nil, fs.ErrNotExist
+}
+
+type cachedDir struct {
+	fs.FileInfo
+	entry []fs.DirEntry
+}
+
+type cachedFileInfo struct {
+	fs.FileInfo
+	hdr    *tar.Header
+	sys    interface{}
+	md5sum []byte
+}
+
+// Ensure that cachedFileInfo implements the CacheHasherFileInfoSum interface.
+var _ util.CacheHasherFileInfoSum = &cachedFileInfo{}
+
+func newCachedFileInfo(hdr *tar.Header) *cachedFileInfo {
+	fi := hdr.FileInfo()
+	return &cachedFileInfo{
+		FileInfo: fi,
+		hdr:      hdr,
+		sys: &syscall.Stat_t{
+			Uid: uint32(hdr.Uid),
+			Gid: uint32(hdr.Gid),
+		},
+	}
+}
+
+func (cf *cachedFileInfo) Sys() interface{} {
+	logrus.Debugf("imagefs: Sys cached file %s", cf.Name())
+	return cf.sys
+}
+
+func (cf *cachedFileInfo) Stat() (fs.FileInfo, error) {
+	logrus.Debugf("imagefs: Stat cached file %s", cf.Name())
+	return cf, nil
+}
+
+func (cf *cachedFileInfo) Read(p []byte) (n int, err error) {
+	panic("imagefs: Read cached file is not allowed")
+}
```

Should we instead return a sentinel error?
At least, we should document the potential panic with big red flashing warnings.